### PR TITLE
dev/core#1846 Regression (sorta) Hacks to avoid failure on upgrade due to new field not yet being in the DB

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
On attempting to upgrade I found I couldn't get to the upgrade screen add the new serialize field as
I got fatal errors in these 2 places (maybe more to come).

The new field is civicrm_custom_field.serialize

Before
----------------------------------------
Can't reach upgrade page due to fatal errors due to db not being upgraded

After
----------------------------------------
short circuit hacks

Technical Details
----------------------------------------

Comments
----------------------------------------

https://lab.civicrm.org/dev/core/-/issues/1846

### CODE REVIEW

File: CRM/Core/BAO/CustomField.php [hunk 1]
```diff
@@ -392,6 +393,10 @@ public static function &getFields(
         if ($onlyParent) {
           $extends .= " AND $cgTable.extends_entity_column_value IS NULL AND $cgTable.extends_entity_column_id IS NULL ";
         }
+        // Temporary hack - in 5.27 a new field is added to civicrm_custom_field. There is a high
+        // risk this function is called before the upgrade page can be reached and if
+        // so it will potentially result in fatal error.
+        $serializeField = CRM_Core_BAO_Domain::isDBUpdateRequired() ? '' : "$cfTable.serialize,";
```
Comment (seamuslee001) [inline CRM/Core/BAO/CustomField.php:399]: what about using instead @eileenmcnaughton 

```suggestion
        $serializeField = CRM_Core_DAO::checkFieldExists('civicrm_custom_field', 'serialize') ? '' : "$cfTable.serialize,";
```

Comment (eileenmcnaughton) [reply CRM/Core/BAO/CustomField.php:399]: I think the one I used would be cached actually

Comment (eileenmcnaughton) [reply CRM/Core/BAO/CustomField.php:399]: (in general schema queries are more likely to cause locks too I think)

Comment (seamuslee001) [reply CRM/Core/BAO/CustomField.php:399]: My understanding was that the checkFieldExists is what we have been using in these sorts of things

Comment (eileenmcnaughton) [reply CRM/Core/BAO/CustomField.php:399]: @seamuslee001 hmm - mostly in upgrade scripts not in functions that could be hit often, even in upgraded dbs

**Comment by seamuslee001** [comment]
@eileenmcnaughton test fails relate

File: CRM/Core/BAO/CustomField.php (hunk 1 above)
Comment (seamuslee001) [reply CRM/Core/BAO/CustomField.php:399]: @eileenmcnaughton you can see its included in other places e.g. in the ConfigSetting.php and Menu.php https://github.com/civicrm/civicrm-core/search?p=2&q=checkIfFieldExists&unscoped_q=checkIfFieldExists so its not just about upgrade scripts

**Comment by kcristiano** [comment]
I did an `r-run`

- Clean install of WP 5.42
- CiviCRM 5.26.2
- Woo Commerce
- Ran Woo Setup
- Added https://github.com/mecachisenros/woocommerce_civicrm

CiviCRM ran fine.

- Updated code base to 5.27 RC
- Applied PR
- Went to `wp-admin/plugins.php` to check CiviCRM version
- Fatal Error
- went to `/wp-admin`
- Worked well
- went to `/wp-admin/admin.php?page=CiviCRM`
- Fatal error
- Cleared cv cache via `cv flush`
- Admin screens loaded
- Went to CiviCRM
- Received System message that an upgrade was needed
- Ran UI Upgrader
- Success

I am not sure why I had to clear cache.  @eileenmcnaughton how does this compare to when you applied the patch?

I'll build a tarball with the patch and see if that makes a difference.

**Comment by kcristiano** [comment]
Using Tarball I still had fatal errors until I cleared the CiviCRM cache.

File: CRM/Core/BAO/CustomField.php (hunk 1 above)
Comment (colemanw) [reply CRM/Core/BAO/CustomField.php:399]: @seamuslee001 it probably shouldn't be used in those places!

**Comment by colemanw** [comment]
I've pushed up some more nuance to the fix.

You know, we have these handy `<add>` and `<drop>` tags in the xml which are (as far as I can tell) completely unused.
It really wouldn't be that hard to teach the api to skip over fields that don't exist yet when we've got upgrades pending and avoid these hard crashes.

Hang on, PR coming...

**Comment by colemanw** [comment]
#17729

**Comment by seamuslee001** [comment]
@eileenmcnaughton @colemanw I still feel like checkIfFieldExists is the correct way of solving this problem but if your wanting to go down this road then you will need to also fix https://github.com/civicrm/civicrm-core/pull/17722/files#diff-50f6e167c4bfdcf605b5e0617d06b82dR754 as that is causing the tests to fail

File: CRM/Core/BAO/CustomField.php [hunk 2]
```diff
@@ -392,6 +393,10 @@ public static function &getFields(
         if ($onlyParent) {
           $extends .= " AND $cgTable.extends_entity_column_value IS NULL AND $cgTable.extends_entity_column_id IS NULL ";
         }
+        // Temporary hack - in 5.27 a new field is added to civicrm_custom_field. There is a high
+        // risk this function is called before the upgrade page can be reached and if
+        // so it will potentially result in fatal error.
+        $serializeField = CRM_Core_BAO_Domain::isDBVersionAtLeast('5.27.alpha1') ? '' : "$cfTable.serialize,";
```
Comment (seamuslee001) [inline CRM/Core/BAO/CustomField.php:399]: Is this around the wrong way?

File: api/v3/CustomField.php [hunk 3]
```diff
@@ -122,7 +122,7 @@ function civicrm_api3_custom_field_delete($params) {
  * @return array
  */
 function civicrm_api3_custom_field_get($params) {
-  if (($params['legacy_html_type'] ?? TRUE) && !empty($params['return'])) {
+  if (!CRM_Core_BAO_Domain::isDBVersionAtLeast('5.27.alpha1') && ($params['legacy_html_type'] ?? TRUE) && !empty($params['return'])) {
```
Comment (seamuslee001) [inline api/v3/CustomField.php:125]: this also seems to be around the wrong way I think

**Comment by eileenmcnaughton** [comment]
Thanks @seamuslee001 & @colemanw this looks good to me - as long as jenkins gets jiggy with it

File: sql/test_data_second_domain.mysql [hunk 4]
```diff
@@ -963,4 +963,4 @@ INSERT INTO civicrm_navigation
 VALUES
     ( @domainID, CONCAT('civicrm/report/instance/', @instanceID,'&reset=1'), 'Mailing Detail Report', 'Mailing Detail Report', 'administer CiviMail', 'OR', @reportlastID, '1', NULL, @instanceID+2 );
 UPDATE civicrm_report_instance SET navigation_id = LAST_INSERT_ID() WHERE id = @instanceID;
-UPDATE civicrm_domain SET version = '4.6.alpha1';
+UPDATE civicrm_domain SET version = '5.27.alpha1';
```
Comment (colemanw) [inline sql/test_data_second_domain.mysql:966]: @seamuslee001 I tackled this in https://github.com/civicrm/civicrm-core/pull/17729/commits/e37554a6366a6470ba27fe1eb13fa055a5af57e0

Comment (colemanw) [reply sql/test_data_second_domain.mysql:966]: @seamuslee001 and @totten tackled it in #17734

**Comment by seamuslee001** [comment]
I'm going to merge this now that Tests have passed and we can revisit the domain version in tests thing in master

**Comment by eileenmcnaughton** [comment]
@seamuslee001 this is definitely better - I wonder if the clear cache thing is still an issue & if there is anything we can do

I guess this is rc now so we have time to re-test

**Comment by eileenmcnaughton** [comment]
No that was wrong  = this is about to become the release

**Comment by shaneonabike** [comment]
When I performed an upgrade I can see that this (I beleive is still happening) at least I think. For the record, I am migrating from 26 to 27.2.

I'm getting:

`WD civicrm: $Fatal Error Details = Array ( [callback] => Array ( [0] => CRM_Core_Error [1] => handle ) [code] => -19 [message] => DB Error: no such field [mode] => 16 [debug_info] => SELECT civicrm_custom_field_en_CA.id as civicrm_custom_field_id, civicrm_custom_field_en_CA.name as civicrm_custom_field_name, civicrm_custom_field_en_CA.label as civicrm_custom_field_label, civicrm_custom_field_en_CA.column_name as civicrm_custom_field_column_name, civicrm_custom_field_en_CA.data_type as civicrm_custom_field_data_type, civicrm_custom_field_en_CA.html_type as civicrm_custom_field_html_type, civicrm_custom_field_en_CA.default_value as civicrm_custom_field_default_value, civicrm_custom_field_en_CA.attributes as civicrm_custom_field_attributes, civicrm_custom_field_en_CA.is_required as civicrm_custom_field_is_required, civicrm_custom_field_en_CA.is_view as civicrm_custom_field_is_view, civicrm_custom_field_en_CA.help_pre as civicrm_custom_field_help_pre, civicrm_custom_field_en_CA.help_post as civicrm_custom_field_help_post, civicrm_custom_field_en_CA.options_per_line as civicrm_custom_field_options_per_line, civicrm_custom_field_en_CA.start_date_years as civicrm_custom_field_start_date_years, civicrm_custom_field_en_CA.end_date_years as civicrm_custom_field_end_date_years, civicrm_custom_field_en_CA.date_format as civicrm_custom_field_date_format, civicrm_custom_field_en_CA.time_format as civicrm_custom_field_time_format, civicrm_custom_field_en_CA.option_group_id as civicrm_custom_field_option_group_id, civicrm_custom_field_en_CA.in_selector as civicrm_custom_field_in_selector, civicrm_custom_field_en_CA.serialize as civicrm_custom_field_serialize, civicrm_custom_group_en_CA.id as civicrm_custom_group_id, civicrm_custom_group_en_CA.name as civicrm_custom_group_name, civicrm_custom_group_en_CA.table_name as civicrm_custom_group_table_name, civicrm_custom_group_en_CA.title as civicrm_custom_group_title, civicrm_custom_group_en_CA.help_pre as civicrm_custom_group_help_pre, civicrm_custom_group_en_CA.help_post as civicrm_custom_group_help_post, civicrm_custom_group_en_CA.collapse_display as civicrm_custom_group_collapse_display, civicrm_custom_group_en_CA.style as civicrm_custom_group_style, civicrm_custom_group_en_CA.is_multiple as civicrm_custom_group_is_multiple, civicrm_custom_group_en_CA.extends as civicrm_custom_group_extends, civicrm_custom_group_en_CA.extends_entity_column_id as civicrm_custom_group_extends_entity_column_id, civicrm_custom_group_en_CA.extends_entity_column_value as civicrm_custom_group_extends_entity_column_value, civicrm_custom_group_en_CA.max_multiple as civicrm_custom_group_max_multiple, civicrm_custom_group_en_CA.is_public as civicrm_custom_group_is_public FROM civicrm_custom_group_en_CA LEFT JOIN civicrm_custom_field_en_CA ON (civicrm_custom_field_en_CA.custom_group_id = civicrm_custom_group_en_CA.id) WHERE civicrm_custom_group_en_CA.is_active = 1 AND civicrm_custom_field_en_CA.is_active = 1 AND civicrm_custom_group_en_CA.extends IN ('Contribution') AND (((civicrm_custom_group_en_CA.extends_entity_column_value LIKE "%1%")) OR civicrm_custom_group_en_CA.extends_entity_column_value IS NULL ) AND civicrm_custom_group_en_CA.id = 1 ORDER BY civicrm_custom_group_en_CA.weight, civicrm_custom_group_en_CA.title, civicrm_custom_field_en_CA.weight,`

I can paste the entire output but I didn't want to flood the ticket.

**Comment by colemanw** [comment]
@shaneonabike was this crash before or after the upgrade ran? Or during?

**Comment by shaneonabike** [comment]
@colemanw I'm using Aegir so this happened during the update process. Everything was installed properly but then during the update this occurred. Then obviously the site rolled itself back afterwards. I'm happy to provide more debugging if needed.

**Comment by colemanw** [comment]
This appears to be a multilingual query but I can't find it in the code by grepping for pieces of it (probably because it was generated). Did you get a backtrace?

**Comment by shaneonabike** [comment]
Yep!

`WD civicrm: $backTrace = #0 /pressflow/sites/all/modules/civicrm/CRM/Core/Error.php(205): CRM_Core_Error::backtrace("backTrace", TRUE) #1 /pressflow/sites/all/modules/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php(922): CRM_Core_Error::handle(Object(DB_Error)) #2 /pressflow/sites/all/modules/civicrm/packages/DB.php(998): PEAR_Error->__construct("DB Error: no such field", -19, 16, (Array:2), "SELECT civicrm_custom_field_en_CA.id as civicrm_custom_field_id, civicrm_cust...") #3 /pressflow/sites/all/modules/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php(575): DB_Error->__construct(-19, 16, (Array:2), "SELECT civicrm_custom_field_en_CA.id as civicrm_custom_field_id, civicrm_cust...") #4 /pressflow/sites/all/modules/civicrm/vendor/pear/pear-core-minimal/src/PEAR.php(223): PEAR->_raiseError(Object(DB_mysqli), NULL, -19, 16, (Array:2), "SELECT civicrm_custom_field_en_CA.id as civicrm_custom_field_id, civicrm_cust...", "DB_Error", TRUE) #5 /pressflow/sites/all/modules/civicrm/packages/DB/common.php(1925): PEAR->__call("raiseError", (Array:7)) #6 /pressflow/sites/all/modules/civicrm/packages/DB/mysqli.php(935): DB_common->raiseError(-19, NULL, NULL, "SELECT civicrm_custom_field_en_CA.id as civicrm_custom_field_id, civicrm_cust...", "1054 ** Unknown column 'civicrm_custom_field_en_CA.serialize' in 'field list'") #7 /pressflow/sites/all/modules/civicrm/packages/DB/mysqli.php(405): DB_mysqli->mysqliRaiseError() #8 /pressflow/sites/all/modules/civicrm/packages/DB/common.php(1231): DB_mysqli->simpleQuery("SELECT civicrm_custom_field_en_CA.id as civicrm_custom_field_id, civicrm_cust...") #9 /pressflow/sites/all/modules/civicrm/packages/DB/DataObject.php(2696): DB_common->query("SELECT civicrm_custom_field_en_CA.id as civicrm_custom_field_id, civicrm_cust...") #10 /pressflow/sites/all/modules/civicrm/packages/DB/DataObject.php(1829): DB_DataObject->_query("SELECT civicrm_custom_field_en_CA.id as civicrm_custom_field_id, civicrm_cust...") #11 /pressflow/sites/all/modules/civicrm/CRM/Core/DAO.php(421): DB_DataObject->query("SELECT civicrm_custom_field_en_CA.id as civicrm_custom_field_id, civicrm_cust...") #12 /pressflow/sites/all/modules/civicrm/CRM/Core/DAO.php(1473): CRM_Core_DAO->query("SELECT civicrm_custom_field_en_CA.id as civicrm_custom_field_id, civicrm_cust...", TRUE) #13 /pressflow/sites/all/modules/civicrm/CRM/Core/BAO/CustomGroup.php(2136): CRM_Core_DAO::executeQuery("SELECT civicrm_custom_field.id as civicrm_custom_field_id, civicrm_custom_fie...", (Array:1)) #14 /pressflow/sites/all/modules/civicrm/CRM/Core/BAO/CustomGroup.php(553): CRM_Core_BAO_CustomGroup::buildGroupTree("Contribution", (Array:2), (Array:1), "SELECT civicrm_custom_field.id as civicrm_custom_field_id, civicrm_custom_fie...", (Array:1), "1") #15 /pressflow/sites/all/modules/civicrm/drupal/modules/views/civicrm.views.inc(334): CRM_Core_BAO_CustomGroup::getTree("Contribution", (Array:2), NULL, "1", (Array:1), NULL, TRUE, NULL, FALSE, FALSE) #16 /pressflow/sites/all/modules/civicrm/drupal/modules/views/components/civicrm.core.inc(3213): civicrm_views_custom_data_cache((Array:254), "Contribution", "1", "1") #17 /pressflow/sites/all/modules/civicrm/drupal/modules/views/civicrm.views.inc(87): _civicrm_core_data((Array:254), (Array:5)) #18 /pressflow/sites/all/modules/contrib/views/includes/cache.inc(95): civicrm_views_data_alter((Array:254)) #19 /pressflow/sites/all/modules/contrib/views/includes/cache.inc(77): _views_fetch_data_build() #20 /pressflow/sites/all/modules/contrib/views/views.module(1420): _views_fetch_data(NULL, TRUE, FALSE) #21 /pressflow/sites/all/modules/contrib/entity/views/entity.views.inc(320): views_fetch_data() #22 /pressflow/sites/all/modules/contrib/views/includes/plugins.inc(417): entity_views_plugins() #23 /pressflow/sites/all/modules/contrib/views/includes/cache.inc(148): views_discover_plugins() #24 /pressflow/sites/all/modules/contrib/views/views.module(1428): _views_fetch_plugin_data(NULL, NULL, FALSE) #25 /pressflow/sites/all/modules/contrib/views/views.module(152): views_fetch_plugin_data() #26 /pressflow/includes/theme.inc(536): views_theme((Array:425), "module", "views", "sites/all/modules/contrib/views") #27 /pressflow/includes/theme.inc(689): _theme_process_registry((Array:425), "views", "module", "views", "sites/all/modules/contrib/views") #28 /pressflow/includes/theme.inc(325): _theme_build_registry(Object(stdClass), (Array:1), "phptemplate") #29 /pressflow/includes/theme.inc(277): _theme_load_registry(Object(stdClass), (Array:1), "phptemplate") #30 /pressflow/includes/theme.inc(438): theme_get_registry() #31 /pressflow/includes/theme.inc(432): ThemeRegistry->resolveCacheMiss("icon") #32 /pressflow/includes/theme.inc(1055): ThemeRegistry->offsetGet("icon") #33 /pressflow/sites/website.org/modules/contrib/icon/modules/icon_filter/icon_filter.module(43): theme("icon", (Array:2)) #34 /pressflow/modules/filter/filter.module(806): _icon_filter("CALACS facilitates networking and information exchange among persons, in ...", Object(stdClass), Object(stdClass), "und", FALSE, "") #35 /pressflow/modules/field/modules/text/text.module(321): check_markup("CALACS facilitates networking and information exchange among persons, in ...", "full_html", "und") #36 /pressflow/modules/field/modules/text/text.module(159): _text_sanitize((Array:13), "und", (Array:3), "value") #37 /pressflow/modules/field/field.attach.inc(344): text_field_load("node", (Array:1), (Array:16), (Array:1), "und", (Array:1), "FIELD_LOAD_CURRENT", NULL) #38 /pressflow/modules/field/field.attach.inc(714): _field_invoke_multiple("load", "node", (Array:1), "FIELD_LOAD_CURRENT", NULL, (Array:3)) #39 /pressflow/includes/entity.inc(354): field_attach_load("node", (Array:1)) #40 /pressflow/modules/node/node.module(4176): DrupalDefaultEntityController->attachLoad((Array:1), FALSE) #41 /pressflow/sites/website.org/modules/contrib/entitycache/includes/entitycache.entitycachecontrollerhelper.inc(150): NodeController->attachLoad((Array:1), FALSE) #42 /pressflow/sites/website.org/modules/contrib/entitycache/includes/entitycache.node.inc(25): EntityCacheControllerHelper::entityCacheLoad(Object(EntityCacheNodeController), (Array:1), (Array:0)) #43 /pressflow/includes/common.inc(8107): EntityCacheNodeController->load((Array:1), (Array:0)) #44 /pressflow/modules/node/node.module(947): entity_load("node", (Array:1), (Array:0), FALSE) #45 /pressflow/modules/node/node.module(966): node_load_multiple((Array:1), (Array:0), FALSE) #46 /pressflow/includes/menu.inc(604): node_load("27") #47 /pressflow/includes/menu.inc(772): _menu_load_objects((Array:23), (Array:2)) #48 /pressflow/includes/menu.inc(482): _menu_translate((Array:23), (Array:2)) #49 /pressflow/includes/menu.inc(1767): menu_get_item() #50 /pressflow/includes/menu.inc(1782): menu_get_custom_theme(TRUE) #51 /pressflow/includes/common.inc(5394): menu_set_custom_theme() #52 /pressflow/includes/bootstrap.inc(2566): _drupal_bootstrap_full() #53 phar:///usr/local/bin/drush/lib/Drush/Boot/DrupalBoot7.php(91): drupal_bootstrap(7) #54 phar:///usr/local/bin/drush/includes/bootstrap.inc(354): Drush\Boot\DrupalBoot7->bootstrap_drupal_full() #55 /usr/share/drush/commands/provision/platform/verify.provision.inc(22): drush_bootstrap(5) #56 phar:///usr/local/bin/drush/includes/command.inc(422): drush_provision_drupal_provision_verify_validate() #57 phar:///usr/local/bin/drush/includes/command.inc(231): _drush_invoke_hooks((Array:37), (Array:0)) #58 phar:///usr/local/bin/drush/includes/command.inc(199): drush_command() #59 phar:///usr/local/bin/drush/lib/Drush/Boot/BaseBoot.php(67): drush_dispatch((Array:37)) #60 phar:///usr/local/bin/drush/includes/preflight.inc(66): Drush\Boot\BaseBoot->bootstrap_and_dispatch() #61 phar:///usr/local/bin/drush/includes/startup.inc(462): drush_main() #62 phar:///usr/local/bin/drush/includes/startup.inc(369): drush_run_main(FALSE, "/", "Phar detected. Proceeding to drush_main().") #63 phar:///usr/local/bin/drush/drush(114): drush_startup((Array:5)) #64 /usr/local/bin/drush(10): require("phar:///usr/local/bin/drush/drush") #65 {main}`

**Comment by shaneonabike** [comment]
Obviously I obfuscated a bit of the paths.

**Comment by colemanw** [comment]
@shaneonabike ok looks like the query was generated by `CRM_Core_BAO_CustomGroup::getTree()`. I've made a PR; want to try manually patching with  #17875 to see if it solves your problem?

**Comment by totten** [comment]
@shaneonabike @colemanw - Thanks for the backtrace! It also helped with updating https://lab.civicrm.org/dev/core/-/issues/1850

**Pull request merged**